### DISCUSSION TITLE
Qt6 Support

### DISCUSSION THREAD

**Discussion by nullobsi** [tsujan/Kvantum#596] [General]
now that qt6 is released, it would be awesome for Kvantum to support qt6! locally I have gotten it to work by modifying the source code however it was very hacky and involved editing the cmakelists in a not good way

the edits I had to make were:
- remove Qt5X11Extras and instead use xlib directly (hacky?)
- upgrade cmakelists to use qt6
- fix some deprecreated/removed functions and properties
- fix qhash for WindowManager::ExceptionId
- update some includes

i don't have very much experience with c++ but I hope this helps! the fork I made with my changes can be found at https://github.com/nullobsi/Kvantum

**Comment by tsujan** [comment]
I checked Qt6.0.0 the same day it was released in Arch and compiled Kvantum against it after some sacrifices and a few changes to the code.

Qt6.0.0 is far from being ready for real use. They shouldn't have pretended that it's a stable release...

**Comment by tsujan** [comment]
OpenSUSE seems to be aware of the Qt6 mess, trying to compensate for it: https://build.opensuse.org/package/show/home:jonaski:qt6/qt6-x11extras

Kvantum aside, I've encountered so many important Qt5 methods, whose future is suspicious in Qt6. It isn't clear what plans Qt devs have for Qt but if they want to "deprecate" all methods in [core5compat](https://doc.qt.io/qt-6/qtcore5-index.html), they should first replace them and tell users about those replacements.

**Comment by NickLion** [comment]
Hi, I agree that Qt 6.0 is basically not a stable release but more like a stable playground for porting to the future stable Qt (which expected to be Qt 6.2).

But what exactly is missing from core5compat ? Replacements were noted in blog (IMHO, it's better them to be placed in documentation, but still):
https://www.qt.io/blog/porting-from-qt-5-to-qt-6-using-qt5compat-library

**Reply by tsujan** [reply]
> But what exactly is missing from core5compat ?

To say nothing of `QX11Info::display()` (which isn't related to core5compat but is really needed), I saw some other methods were missing from `qt6-5compat` when I tried to port another app (FeatherPad) to Qt6.

The question is that, if the remaining classes and methods are *only* for compatibility with Qt5, what does Qt6 have to offer instead of them? I don't like my apps to be in an transitory state. `qt6-5compat` isn't a real solution.

**Reply by NickLion** [reply]
Yeah, Extras are definitely a WIP ([QTBUG-83251](https://bugreports.qt.io/browse/QTBUG-83251), [QTBUG-80233](https://bugreports.qt.io/browse/QTBUG-80233)), I was wondered just about qt5compat.

**Reply by tsujan** [reply]
> I was wondered just about qt5compat.

I tried some Qt6 replacements for qt5compat classes (in another app) and they worked fine :)

Anyway, Qt6 seems to be more backward-incompatible than Qt5. That isn't a problem in itself because improvement may require backward-incompatibility but old codes will stop working. IMHO, a real Qt6 port of an app should avoid qt5compat.

**Comment by tsujan** [comment]
@NickLion 
The table at https://www.qt.io/blog/porting-from-qt-5-to-qt-6-using-qt5compat-library may be the answer. Thanks for the link!

Since I hate `core5compat`, I should wait to see what Qt6 will offer (in 6.1 or 6.2). Luckily, Kvantum won't need `core5compat` but I think parts of other codes will need to be rewritten for Qt6.

**Comment by shanoaice** [comment]
Seemed like that Qt6.2 will be soon in release......any plan to port Kvantum to qt6.2 for now?

**Reply by tsujan** [reply]
They say Qt 6.2 has x11extras. If so,  yes. Although there's no rush.

**Reply by tsujan** [reply]
> Seemed like that Qt6.2 will be soon in release......any plan to port Kvantum to qt6.2 for now?

In addition to `x11extras`, `KWindowSystem` needs to be ported to Qt6, get stable and come to distros like Arch, Manjaro and Tumbleweed (to say nothing of Debian). So, we still have to wait. Otherwise, the porting is straightforward.

**Comment by tsujan** [comment]
Qt 6.2.0 has `QNativeInterface::QX11Application`. So, there's no need to `x11extras` anymore. But a stable version of `KF6WindowSystem` doesn't exist yet.

Actually, I made a Qt6 patch ready and excluded `KWindowSystem` in it. Kvantum was compiled and worked fine against Qt6. The only missing functionality was blurring (provided by `KWindowSystem`).

If Qt6 becomes the default Qt version in Arch and Manjaro Testing, I'll modify and apply the patch. Compilation against Qt5 will still be possible but without Kvantum Manager (and `kvantumpreview`), which will be for Qt6.

EDIT: I added experimental Qt6 support. See https://github.com/tsujan/Kvantum/blob/master/Kvantum/INSTALL.md#with-cmake → NOTE. Of course, it'll change.

**Comment by enkeyz** [comment]
Any word on the progress? Arch and Manjaro going with Qt6 right now, and themes(arc-dark from papyrus team) applied with kvantum to Qt6 apps - qBittorrent for example - are broken.

![image](https://user-images.githubusercontent.com/17646853/151348480-8c829358-b335-48f5-96f7-ee605c4cab02.png)

**Reply by tsujan** [reply]
> Any word on the progress?

It was done more than a month ago; Kvantum 1.0.0 can be compiled against Qt6. Read https://github.com/tsujan/Kvantum/blob/master/Kvantum/INSTALL.md#with-cmake → NOTE

If you compile it, for your Qt6 apps to see it, you'll also need a Qt6 platform plugin, but KDE and LXQt still aren't ported to Qt6. You could use temporary methods that are explained in `INSTALL` but icons will need a good Qt6 icon engine, and only KDE and LXQt could provide it when they're ported to Qt6.

qBittorrent with Qt6 + Kvantum + LXQt platform plugin that's partially ported to Qt6 by me, under LXQt:

![qbittorrent](https://user-images.githubusercontent.com/981076/151361974-d5630dc5-c0e1-4bfd-8b97-ee1087f8a6cf.png)

As you see, KWin's blurring still doesn't work because `KF6` isn't released yet.

**Reply by enkeyz** [reply]
So it's a waiting game. Thanks for the reply!

**Reply by tsujan** [reply]
> So it's a waiting game.

Yes, for me too, because I need `KF6` to enable the blurring support in Kvantum + Qt6.

As soon as Qt6 KDE comes to Manjaro, I'll set the default Qt version to 6 in Kvantum and make a release. Package maintainers will need to rebuild Qt5 packages, so that both Qt5 and Qt6 packages could be installed.
